File: src/domain/entities/learning_content.py
"""
Domain entities for enriched learning content.

This module defines the data models for structured educational content
including lessons, concept cards, code examples, and knowledge checks.
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional, Dict, Any
from uuid import UUID, uuid4


class ContentSectionType(str, Enum):
    """Types of content sections within a lesson."""
    TEXT = "text"
    CONCEPT_CARD = "concept-card"
    CODE_EXAMPLE = "code-example"
    KNOWLEDGE_CHECK = "knowledge-check"
    DIAGRAM = "diagram"


class KnowledgeCheckType(str, Enum):
    """Types of knowledge check questions."""
    MULTIPLE_CHOICE = "multiple-choice"
    FILL_BLANK = "fill-blank"
    CODE_COMPLETION = "code-completion"
    TRUE_FALSE = "true-false"


class DiagramType(str, Enum):
    """Types of Mermaid diagrams."""
    FLOWCHART = "flowchart"
    SEQUENCE = "sequence"
    CLASS = "class"
    STATE = "state"
    ER = "er"


class ResourceCategory(str, Enum):
    """Categories for external resources."""
    ESSENTIAL = "essential"
    RECOMMENDED = "recommended"
    DEEP_DIVE = "deep-dive"


class ResourceType(str, Enum):
    """Types of external resources."""
    DOCUMENTATION = "documentation"
    TUTORIAL = "tutorial"
    VIDEO = "video"
    ARTICLE = "article"
    REFERENCE = "reference"


class ProgrammingLanguage(str, Enum):
    """Supported programming languages for code examples."""
    JAVASCRIPT = "javascript"
    TYPESCRIPT = "typescript"
    PYTHON = "python"
    JAVA = "java"


@dataclass
class LessonMetadata:
    """Metadata for a lesson."""
    estimated_minutes: int
    difficulty: str  # beginner, intermediate, advanced
    prerequisites: List[str] = field(default_factory=list)
    technology: str = ""
    last_updated: str = ""


@dataclass
class TextBlock:
    """A block of text content."""
    content: str
    format: str = "markdown"  # markdown or html


@dataclass
class Mistake:
    """A common mistake with correction."""
    description: str
    example: str
    correction: str


@dataclass
class UseCase:
    """A practical use case for a concept."""
    scenario: str
    example: str
    benefit: str


@dataclass
class Analogy:
    """A real-world analogy for a concept."""
    title: str
    description: str
    mapping: Dict[str, str] = field(default_factory=dict)  # concept term -> analogy term


@dataclass
class CodeSnippet:
    """A small code snippet within a concept card."""
    language: str
    code: str
    description: str = ""


@dataclass
class MermaidDiagram:
    """A Mermaid diagram for visualization."""
    type: DiagramType
    code: str
    caption: str
    alt_text: str  # Accessibility


@dataclass
class ConceptCard:
    """A concept card with multiple explanation styles."""
    id: str
    concept_name: str
    primary_explanation: str
    analogy: Analogy
    diagram: Optional[MermaidDiagram] = None
    alternative_explanations: List[str] = field(default_factory=list)
    common_mistakes: List[Mistake] = field(default_factory=list)
    when_to_use: List[UseCase] = field(default_factory=list)
    code_snippet: Optional[CodeSnippet] = None


@dataclass
class TestCase:
    """A test case for code examples."""
    input: str
    expected_output: str
    description: str


@dataclass
class CodeExample:
    """An interactive code example."""
    id: str
    title: str
    description: str
    language: ProgrammingLanguage
    starter_code: str
    solution_code: str
    test_cases: List[TestCase] = field(default_factory=list)
    hints: List[str] = field(default_factory=list)
    is_editable: bool = True
    expected_output: Optional[str] = None


@dataclass
class Option:
    """An option for multiple choice questions."""
    id: str
    text: str
    is_correct: bool
    feedback: str


@dataclass
class KnowledgeCheck:
    """A knowledge check question."""
    id: str
    question: str
    type: KnowledgeCheckType
    options: List[Option] = field(default_factory=list)
    correct_answer: str = ""  # Can be string or JSON array for multiple answers
    explanation: str = ""
    hint: str = ""
    related_concept_id: Optional[str] = None
    difficulty: int = 1  # 1-5


@dataclass
class ContentSection:
    """A section of content within a lesson."""
    id: str
    type: ContentSectionType
    order: int
    content: Any  # TextBlock, ConceptCard, CodeExample, KnowledgeCheck, or MermaidDiagram
    completion_required: bool = True


@dataclass
class Resource:
    """An external learning resource."""
    id: str
    title: str
    url: str
    type: ResourceType
    category: ResourceCategory
    estimated_minutes: int
    verified: bool = False
    last_verified: str = ""


@dataclass
class StructuredLesson:
    """A complete structured lesson."""
    id: str
    title: str
    topic: str
    metadata: LessonMetadata
    objectives: List[str]
    sections: List[ContentSection]
    key_takeaways: List[str]
    related_resources: List[Resource] = field(default_factory=list)
    version: str = "1.0.0"
    
    def get_total_sections(self) -> int:
        """Get total number of sections."""
        return len(self.sections)
    
    def get_knowledge_checks(self) -> List[KnowledgeCheck]:
        """Get all knowledge checks in the lesson."""
        return [
            s.content for s in self.sections 
            if s.type == ContentSectionType.KNOWLEDGE_CHECK
        ]
    
    def get_code_examples(self) -> List[CodeExample]:
        """Get all code examples in the lesson."""
        return [
            s.content for s in self.sections 
            if s.type == ContentSectionType.CODE_EXAMPLE
        ]
    
    def get_concept_cards(self) -> List[ConceptCard]:
        """Get all concept cards in the lesson."""
        return [
            s.content for s in self.sections 
            if s.type == ContentSectionType.CONCEPT_CARD
        ]


@dataclass
class LessonProgress:
    """User progress through a lesson."""
    user_id: UUID
    lesson_id: UUID
    current_section_id: Optional[str] = None
    completed_sections: List[str] = field(default_factory=list)
    knowledge_check_results: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    scroll_position: int = 0
    time_spent_seconds: int = 0
    completed: bool = False
    completed_at: Optional[datetime] = None
    last_accessed_at: datetime = field(default_factory=datetime.now)
    
    def get_completion_percentage(self, total_sections: int) -> float:
        """Calculate completion percentage."""
        if total_sections == 0:
            return 0.0
        return (len(self.completed_sections) / total_sections) * 100


@dataclass
class UserNote:
    """A user's note or highlight."""
    id: UUID = field(default_factory=uuid4)
    user_id: UUID = field(default_factory=uuid4)
    lesson_id: UUID = field(default_factory=uuid4)
    section_id: Optional[str] = None
    note_type: str = "note"  # "highlight" or "note"
    content: str = ""
    selection_start: Optional[int] = None
    selection_end: Optional[int] = None
    color: Optional[str] = None
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)


@dataclass
class KnowledgeCheckAttemptRecord:
    """Record of a knowledge check attempt."""
    id: UUID = field(default_factory=uuid4)
    user_id: UUID = field(default_factory=uuid4)
    lesson_id: UUID = field(default_factory=uuid4)
    check_id: str = ""
    answer: str = ""
    is_correct: bool = False
    attempt_number: int = 1
    time_taken_seconds: Optional[int] = None
    created_at: datetime = field(default_factory=datetime.now)

File: frontend/src/types/learning-content.ts
/**
 * TypeScript interfaces for enriched learning content.
 * 
 * These types mirror the Python domain entities and add UI-specific properties
 * for managing component state.
 */

// Enums
export type ContentSectionType = 'text' | 'concept-card' | 'code-example' | 'knowledge-check' | 'diagram';
export type KnowledgeCheckType = 'multiple-choice' | 'fill-blank' | 'code-completion' | 'true-false';
export type DiagramType = 'flowchart' | 'sequence' | 'class' | 'state' | 'er';
export type ResourceCategory = 'essential' | 'recommended' | 'deep-dive';
export type ResourceType = 'documentation' | 'tutorial' | 'video' | 'article' | 'reference';
export type ProgrammingLanguage = 'javascript' | 'typescript' | 'python' | 'java';
export type SkillLevel = 'beginner' | 'intermediate' | 'advanced';
export type NoteType = 'highlight' | 'note';
export type HighlightColor = 'yellow' | 'green' | 'blue' | 'pink';

// Metadata
export interface LessonMetadata {
  estimatedMinutes: number;
  difficulty: SkillLevel;
  prerequisites: string[];
  technology: string;
  lastUpdated: string;
}

// Text content
export interface TextBlock {
  content: string;
  format: 'markdown' | 'html';
}

// Concept Card components
export interface Mistake {
  description: string;
  example: string;
  correction: string;
}

export interface UseCase {
  scenario: string;
  example: string;
  benefit: string;
}

export interface Analogy {
  title: string;
  description: string;
  mapping: Record<string, string>; // concept term -> analogy term
}

export interface CodeSnippet {
  language: string;
  code: string;
  description?: string;
}

export interface MermaidDiagram {
  type: DiagramType;
  code: string;
  caption: string;
  altText: string; // Accessibility
}

export interface ConceptCard {
  id: string;
  conceptName: string;
  primaryExplanation: string;
  analogy: Analogy;
  diagram?: MermaidDiagram;
  alternativeExplanations: string[];
  commonMistakes: Mistake[];
  whenToUse: UseCase[];
  codeSnippet?: CodeSnippet;
}

// Code Example
export interface TestCase {
  input: string;
  expectedOutput: string;
  description: string;
}

export interface CodeExample {
  id: string;
  title: string;
  description: string;
  language: ProgrammingLanguage;
  starterCode: string;
  solutionCode: string;
  testCases: TestCase[];
  hints: string[];
  isEditable: boolean;
  expectedOutput?: string;
}

// Knowledge Check
export interface Option {
  id: string;
  text: string;
  isCorrect: boolean;
  feedback: string;
}

export interface KnowledgeCheck {
  id: string;
  question: string;
  type: KnowledgeCheckType;
  options: Option[];
  correctAnswer: string | string[];
  explanation: string;
  hint: string;
  relatedConceptId?: string;
  difficulty: number; // 1-5
}

// Content Section
export interface ContentSection {
  id: string;
  type: ContentSectionType;
  order: number;
  content: TextBlock | ConceptCard | CodeExample | KnowledgeCheck | MermaidDiagram;
  completionRequired: boolean;
}

// Resource
export interface Resource {
  id: string;
  title: string;
  url: string;
  type: ResourceType;
  category: ResourceCategory;
  estimatedMinutes: number;
  verified: boolean;
  lastVerified: string;
}

// Structured Lesson
export interface StructuredLesson {
  id: string;
  title: string;
  topic: string;
  metadata: LessonMetadata;
  objectives: string[];
  sections: ContentSection[];
  keyTakeaways: string[];
  relatedResources: Resource[];
  version: string;
}

// Progress tracking
export interface KnowledgeCheckResult {
  checkId: string;
  isCorrect: boolean;
  attempts: number;
  lastAnswer: string;
  completedAt?: string;
}

export interface LessonProgress {
  userId: string;
  lessonId: string;
  currentSectionId?: string;
  completedSections: string[];
  knowledgeCheckResults: Record<string, KnowledgeCheckResult>;
  scrollPosition: number;
  timeSpentSeconds: number;
  completed: boolean;
  completedAt?: string;
  lastAccessedAt: string;
}

// User notes and highlights
export interface UserNote {
  id: string;
  userId: string;
  lessonId: string;
  sectionId?: string;
  noteType: NoteType;
  content: string;
  selectionStart?: number;
  selectionEnd?: number;
  color?: HighlightColor;
  createdAt: string;
  updatedAt: string;
}

// UI State interfaces
export interface ConceptCardUIState extends ConceptCard {
  isExpanded: boolean;
  showAlternativeExplanation: boolean;
  currentAlternativeIndex: number;
}

export interface CodeExampleUIState extends CodeExample {
  currentCode: string;
  isRunning: boolean;
  output?: string;
  error?: string;
  showHints: boolean;
  currentHintIndex: number;
  hasBeenModified: boolean;
}

export interface KnowledgeCheckUIState extends KnowledgeCheck {
  selectedAnswer?: string;
  isSubmitted: boolean;
  isCorrect?: boolean;
  attempts: number;
  showHint: boolean;
  showExplanation: boolean;
}

export interface ContentSectionUIState extends ContentSection {
  isCompleted: boolean;
  isActive: boolean;
  isVisible: boolean;
}

export interface LessonUIState {
  lesson: StructuredLesson;
  sections: ContentSectionUIState[];
  progress: LessonProgress;
  notes: UserNote[];
  isLoading: boolean;
  error?: string;
  showResumePrompt: boolean;
  autoSaveEnabled: boolean;
}

// API Request/Response types
export interface GenerateLessonRequest {
  topic: string;
  taskTitle: string;
  skillLevel: SkillLevel;
  technology?: string;
  requirements?: string[];
}

export interface GenerateLessonResponse {
  lesson: StructuredLesson;
  generated: boolean;
}

export interface SaveProgressRequest {
  lessonId: string;
  currentSectionId?: string;
  completedSections: string[];
  scrollPosition: number;
  timeSpentSeconds: number;
}

export interface SubmitKnowledgeCheckRequest {
  lessonId: string;
  checkId: string;
  answer: string;
  timeTakenSeconds?: number;
}

export interface SubmitKnowledgeCheckResponse {
  isCorrect: boolean;
  feedback: string;
  explanation: string;
  attemptNumber: number;
  shouldReExplain: boolean;
}

export interface ExplainDifferentlyRequest {
  conceptId: string;
  previousExplanations: string[];
}

export interface ExplainDifferentlyResponse {
  explanation: string;
  analogy?: Analogy;
}

export interface CreateNoteRequest {
  lessonId: string;
  sectionId?: string;
  noteType: NoteType;
  content: string;
  selectionStart?: number;
  selectionEnd?: number;
  color?: HighlightColor;
}

export interface ExportNotesResponse {
  markdown: string;
  filename: string;
}

// Code execution types
export interface CodeExecutionRequest {
  code: string;
  language: ProgrammingLanguage;
  testCases?: TestCase[];
}

export interface CodeExecutionResult {
  success: boolean;
  output: string;
  error?: string;
  testResults?: TestCaseResult[];
  executionTimeMs: number;
}

export interface TestCaseResult {
  passed: boolean;
  input: string;
  expectedOutput: string;
  actualOutput: string;
  description: string;
}

// Helper type guards
export function isTextBlock(content: unknown): content is TextBlock {
  return typeof content === 'object' && content !== null && 'content' in content && 'format' in content;
}

export function isConceptCard(content: unknown): content is ConceptCard {
  return typeof content === 'object' && content !== null && 'conceptName' in content && 'primaryExplanation' in content;
}

export function isCodeExample(content: unknown): content is CodeExample {
  return typeof content === 'object' && content !== null && 'starterCode' in content && 'language' in content;
}

export function isKnowledgeCheck(content: unknown): content is KnowledgeCheck {
  return typeof content === 'object' && content !== null && 'question' in content && 'type' in content;
}

export function isMermaidDiagram(content: unknown): content is MermaidDiagram {
  return typeof content === 'object' && content !== null && 'code' in content && 'type' in content && 'altText' in content;
}
